query-service-impl/src/main/java/org/hypertrace/core/query/service/ExecutionContext.java

```diff
@@ -35,14 +39,62 @@
   // while the selectedColumns
   // is a set of column names.
   private final LinkedHashSet<Expression> allSelections;
+  private final Optional<Double> timeSeriesPeriod;
+  private final double timeRangeDuration;
 
   public ExecutionContext(String tenantId, QueryRequest request) {
     this.tenantId = tenantId;
     this.selectedColumns = new LinkedHashSet<>();
     this.allSelections = new LinkedHashSet<>();
+    this.timeSeriesPeriod = setTimeSeriesPeriod(request);
+    this.timeRangeDuration = setTimeRangeDuration(request);
     analyze(request);
   }
 
+  private Optional<Double> setTimeSeriesPeriod(QueryRequest request) {
+    Optional<Double> period = Optional.empty();
+    if (request.getGroupByCount() > 0) {
+      for (Expression expression : request.getGroupByList()) {
+        if (expression.getValueCase() == ValueCase.FUNCTION
+            && expression.getFunction().getFunctionName().equals("dateTimeConvert")) {
+          String periodInIso =
+              expression
+                  .getFunction()
+                  .getArgumentsList()
+                  .get(3)
+                  .getLiteral()
+                  .getValue()
+                  .getString();
+          period = Optional.of((double) isoDurationToSeconds(periodInIso));
+        }
+      }
+    }
+    return period;
+  }
+
+  private double setTimeRangeDuration(QueryRequest request) {
+    Optional<Long> duration = Optional.empty();
+    if (request.getFilter().getChildFilterCount() > 0) {
```

Comment:
We're assuming the filters will be specified at the top level, which I don't think is safe.

Reply:
Continued discussion in another thread.